src/sentry/workflow_engine/processors/data_condition_group.py

```diff
@@ -1,31 +1,182 @@
 import dataclasses
 import logging
-from typing import TypeVar
+from collections.abc import Callable, Iterable
+from typing import ClassVar, NoReturn, TypeVar
 
 import sentry_sdk
 
 from sentry.utils.function_cache import cache_func_for_models
 from sentry.workflow_engine.models import DataCondition, DataConditionGroup
 from sentry.workflow_engine.models.data_condition import is_slow_condition
 from sentry.workflow_engine.processors.data_condition import split_conditions_by_speed
-from sentry.workflow_engine.types import DataConditionResult
+from sentry.workflow_engine.types import ConditionError, DataConditionResult
 from sentry.workflow_engine.utils import scopedstats
 
 logger = logging.getLogger(__name__)
 
 T = TypeVar("T")
 
 
+def _find_error(
+    items: list["TriggerResult"], predicate: Callable[["TriggerResult"], bool]
+) -> ConditionError | None:
+    """Helper to find an error from items matching the predicate."""
+    return next((item.error for item in items if predicate(item)), None)
+
+
+@dataclasses.dataclass(frozen=True)
+class TriggerResult:
```

Comment:
🤔 wonder if we should make `EvaluationResult` more generic; then we could share it in all the places we have evaluations. it'd probably just be taking a generic type here for `ConditionError` and maybe adding a `attributes` or something dict that we can toss debug data into.